opendata_module/metrics_statistics/mongodb_manager.py

```diff
@@ -0,0 +1,260 @@
+""" MongoDB Manager - Opendata Module
+"""
+
+#  The MIT License
+#  Copyright (c) 2021- Nordic Institute for Interoperability Solutions (NIIS)
+#  Copyright (c) 2017-2020 Estonian Information System Authority (RIA)
+#
+#  Permission is hereby granted, free of charge, to any person obtaining a copy
+#  of this software and associated documentation files (the "Software"), to deal
+#  in the Software without restriction, including without limitation the rights
+#  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+#  copies of the Software, and to permit persons to whom the Software is
+#  furnished to do so, subject to the following conditions:
+#
+#  The above copyright notice and this permission notice shall be included in
+#  all copies or substantial portions of the Software.
+#
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+
+import logging
+import urllib.parse
+from datetime import datetime, timedelta
+from typing import Any, Dict, List, Mapping, Optional, TypedDict
+
+from pymongo import MongoClient
+
+# All ts in metrics are in milliseconds.
+METRICS_TS_MULTIPLIER = 1000
+
+
+def metrics_ts(dt: datetime) -> float:
+    return dt.timestamp() * METRICS_TS_MULTIPLIER
+
+
+class RequestsCountData(TypedDict):
+    total_request_count: int
+    previous_year_request_count: int
+    current_year_request_count: int
+    previous_month_request_count: int
+    current_month_request_count: int
+    today_request_count: int
+
+
+class DateTimeRange:
+    def __init__(self, from_dt: datetime, to_dt: Optional[datetime] = None) -> None:
+        """
+            Represents a range of datetime values.
+
+            Parameters:
+                from_ (datetime): The starting datetime of the range.
+                to_ (datetime, optional): The ending datetime of the range. Defaults to None.
+        """
+        self.from_dt = from_dt
+        self.to_dt = to_dt
+
+    def from_ts(self) -> float:
+        return metrics_ts(self.from_dt)
+
+    def to_ts(self) -> float:
+        if not self.to_dt:
+            self.to_dt = datetime.now()
+        return metrics_ts(self.to_dt)
+
+
+class RequestCountDateTimeRange(TypedDict):
+    total_request_count: None
+    previous_year_request_count: DateTimeRange
+    current_year_request_count: DateTimeRange
+    previous_month_request_count: DateTimeRange
+    current_month_request_count: DateTimeRange
+    today_request_count: DateTimeRange
```

Comment:
Optional: it's quite convenient to have typeddicts at the top of file and only then have classes